File: app/src/main/java/com/jann_luellmann/thekenapp/ListFragment.java
package com.jann_luellmann.thekenapp;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.evrencoskun.tableview.TableView;
import com.jann_luellmann.thekenapp.data.model.Bought;
import com.jann_luellmann.thekenapp.data.model.Customer;
import com.jann_luellmann.thekenapp.data.model.Drink;
import com.jann_luellmann.thekenapp.data.model.Event;
import com.jann_luellmann.thekenapp.data.model.relationship.CustomerWithBought;
import com.jann_luellmann.thekenapp.data.model.relationship.EventWithDrinksAndCustomers;
import com.jann_luellmann.thekenapp.data.view_model.relationship.EventWithDrinksAndCustomersViewModel;
import com.jann_luellmann.thekenapp.util.Prefs;
import com.jann_luellmann.thekenapp.view.Cell;
import com.jann_luellmann.thekenapp.view.ColumnHeader;
import com.jann_luellmann.thekenapp.view.ListTableViewAdapter;
import com.jann_luellmann.thekenapp.view.ListTableViewClickListener;
import com.jann_luellmann.thekenapp.view.RowHeader;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import lombok.NoArgsConstructor;

@NoArgsConstructor
public class ListFragment extends Fragment implements EventChangedListener {

    private TableView tableView;
    private ListTableViewAdapter adapter;

    private EventWithDrinksAndCustomersViewModel eventWithDrinksAndCustomersViewModel;

    private List<RowHeader> rowHeaderList = new ArrayList<>();
    private List<ColumnHeader> columnHeaderList = new ArrayList<>();
    private List<List<Cell>> cellList = new ArrayList<>();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_list, container, false);
        tableView = view.findViewById(R.id.tableView);
        tableView.setIgnoreSelectionColors(true);
        tableView.setHasFixedWidth(false);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        this.adapter = new ListTableViewAdapter();
        this.tableView.setAdapter(adapter);
        this.adapter.setAllItems(columnHeaderList, rowHeaderList, cellList);

        long eventId = Prefs.getLong(getContext(), Prefs.CURRENT_EVENT, 1L);
        eventWithDrinksAndCustomersViewModel = ViewModelProviders.of(this).get(EventWithDrinksAndCustomersViewModel.class);
        onEventUpdated(eventId);
    }

    @Override
    public void onEventUpdated(long eventId) {
        if(eventWithDrinksAndCustomersViewModel == null) {
            if(isAdded()) {
                eventWithDrinksAndCustomersViewModel = ViewModelProviders.of(this).get(EventWithDrinksAndCustomersViewModel.class);
            }
            else {
                return;
            }
        }

        eventWithDrinksAndCustomersViewModel.findById(eventId).observe(this, event -> {
            this.columnHeaderList.clear();
            this.rowHeaderList.clear();
            this.cellList.clear();

            setData(event);

            this.adapter.setAllItems(columnHeaderList, rowHeaderList, cellList);
            this.adapter.notifyDataSetChanged();
            this.tableView.setTableViewListener(new ListTableViewClickListener(getContext(), event));
        });
    }

    private void setData(EventWithDrinksAndCustomers event) {
        if(event == null)
            return;

        for (Drink drink : event.getDrinks()) {
            columnHeaderList.add(new ColumnHeader(drink.toString()));
        }

        columnHeaderList.add(new ColumnHeader(getString(R.string.total)));

        float total = 0f;

        Collections.sort(event.getCustomerWithBoughts());
        for (CustomerWithBought customer : event.getCustomerWithBoughts()) {
            rowHeaderList.add(new RowHeader(customer.getCustomer().toString()));

            float sum = 0f;

            List<Cell> row = new ArrayList<>();
            for (Drink drink : event.getDrinks()) {
                boolean isBoughtPresent = false;
                for (Bought bought : customer.getBoughtsByEvent(event.getEvent().getEventId())) {
                    if(bought.getDrinkId() == drink.getDrinkId()) {
                        sum += bought.getAmount() * drink.getPrice();
                        row.add(new Cell(bought.getAmount()));
                        isBoughtPresent = true;
                        break;
                    }
                }
                if(!isBoughtPresent) {
                    row.add(new Cell(0));
                }
            }

            total += sum;
            row.add(new Cell(String.format(Locale.GERMAN, "%.2f€", (sum / 100f))));
            cellList.add(row);
        }

        rowHeaderList.add(new RowHeader(getString(R.string.total)));

        // Add total sum of all customers
        List<Cell> lastRow = new ArrayList<>();
        for (int i = 0; i < event.getDrinks().size(); i++) {
            lastRow.add(new Cell(""));
        }
        lastRow.add(new Cell(String.format(Locale.GERMAN, "%.2f€", (total / 100f))));

        cellList.add(lastRow);
    }
}

File: app/src/main/java/com/jann_luellmann/thekenapp/ListFragment.kt
package com.jann_luellmann.thekenapp

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import com.jann_luellmann.thekenapp.data.model.Customer
import com.jann_luellmann.thekenapp.data.model.relationship.CustomerEventWithBoughtDrinks
import com.jann_luellmann.thekenapp.data.view_model.relationship.EventAndCustomerWithDrinksViewModel
import com.jann_luellmann.thekenapp.data.view_model.relationship.EventWithDrinksAndCustomersViewModel
import com.jann_luellmann.thekenapp.databinding.FragmentListBinding
import com.jann_luellmann.thekenapp.dialog.AddDrinkDialogFragment
import com.jann_luellmann.thekenapp.view.*
import java.util.*


class ListFragment(
    private val eventId: LiveData<Long>,
    private val eventWithDrinksAndCustomersViewModel: EventWithDrinksAndCustomersViewModel,
    private val eventAndCustomerWithDrinksViewModel: EventAndCustomerWithDrinksViewModel
) : Fragment(), EventChangedListener, OnCustomerClickedListener {

    private lateinit var binding: FragmentListBinding

    private var adapter: ListTableViewAdapter = ListTableViewAdapter()

    private val rowHeaderList: MutableList<RowHeader?> = mutableListOf()
    private val columnHeaderList: MutableList<ColumnHeader?> = mutableListOf()
    private val cellList: MutableList<List<Cell?>> = mutableListOf()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view: View = inflater.inflate(R.layout.fragment_list, container, false)
        binding = FragmentListBinding.bind(view)

        binding.tableView.isIgnoreSelectionColors = true
        binding.tableView.setHasFixedWidth(false)
        adapter.tableView = binding.tableView

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.tableView.setAdapter(adapter)
        adapter.setAllItems(columnHeaderList, rowHeaderList, cellList)

        eventId.observe(viewLifecycleOwner) {
            onEventUpdated(eventId)
        }

        binding.tableView.tableViewListener = ListTableViewClickListener()
    }

    override fun onEventUpdated(updatedId: LiveData<Long>) {
        val eventWithCustomersDrinks = Transformations.switchMap(updatedId) {
            eventWithDrinksAndCustomersViewModel.findById(it)
        }

        eventWithCustomersDrinks.observe(viewLifecycleOwner) { event ->
                if (event == null) {
                    binding.noEventsMessage.visibility = View.VISIBLE
                    return@observe
                }

                binding.noEventsMessage.visibility = View.GONE

                columnHeaderList.clear()
                event.drinks.sortedBy { it.name }.forEach {
                    columnHeaderList.add(ColumnHeader(it.toString()))
                }
                columnHeaderList.add(ColumnHeader(getString(R.string.total)))
            }

        val eventAndCustomerWithDrinks = Transformations.switchMap(updatedId) {
            eventAndCustomerWithDrinksViewModel.findAllByEvent(it)
        }

        eventAndCustomerWithDrinks.observe(viewLifecycleOwner) {
            rowHeaderList.clear()
            cellList.clear()
            setData(it)
            adapter.setAllItems(columnHeaderList, rowHeaderList, cellList)
            adapter.notifyDataSetChanged()
        }
    }

    private fun setData(event: List<CustomerEventWithBoughtDrinks>) {
        var total = 0f
        event
            .sortedBy { it.customer.name }
            .groupBy { it.customer }
            .forEach { (customer, entry) ->
                rowHeaderList.add(RowHeaderCustomer(customer, this))
                var sum = 0f
                val row: MutableList<Cell> = mutableListOf()

                entry.sortedBy { it.drink.name }.forEachIndexed { index, it ->
                    if (index < columnHeaderList.size) {
                        sum += it.amount * it.drink.price
                        row.add(
                            Cell(

                                if (columnHeaderList[index]?.data.toString() == it.drink.toString())
                                    it.amount
                                else
                                    "WRONG"
                            )
                        )
                    }
                }
                total += sum
                row.add(Cell(String.format(Locale.GERMAN, "%.2f€", sum / 100f)))
                cellList.add(row)
            }

        if (event.isNotEmpty()) {
            rowHeaderList.add(RowHeader(getString(R.string.total)))

            // Add last row with total sum of all customers
            cellList.add(mutableListOf<Cell>().apply {
                repeat(columnHeaderList.size - 1) { add(Cell("")) }
                add(Cell(String.format(Locale.GERMAN, "%.2f€", total / 100f)))
            })
        }
    }

    override fun onCustomerClicked(customer: Customer) {
        context?.let {
            eventId.observe(viewLifecycleOwner) {
                val dialog = AddDrinkDialogFragment(
                    eventAndCustomerWithDrinksViewModel,
                    it,
                    customer.customerId
                )
                dialog.show(parentFragmentManager, "AddDrink")
            }
        }
    }
}